lib/api/core/indexCache.js

```diff
@@ -32,113 +33,135 @@ const
-              })
-              .then(defaultMappings => {
-                // We are working on the same index so the returned mappings are all the same
-                this.defaultMappings[index] = defaultMappings[0];
-              })
-          );
-        }
+    const promises = [];
 
-        return Bluebird.all(promises);
-      })
-      .then(() => internalEngine.listAliases())
-      .then(aliases => {
-        for (const alias of Object.keys(aliases)) {
-          const original = aliases[alias];
-          this.indexes[alias] = this.indexes[original];
-          this.defaultMappings[alias] = this.defaultMappings[original];
-        }
-      });
+    promises.push(
+      this.internalStorage.listIndexes()
+        .then(indexes => this._addCollections(indexes, 'internal')));
+
+    promises.push(
+      this.publicStorage.listIndexes()
+        .then(indexes => this._addCollections(indexes, 'public')));
+
+    return Bluebird.all(promises);
   }
 
-  add (index, collection, notify = true) {
+  /**
+   * Lists collections from all indexes passed in argument and add them to the cache
+   *
+   * @param {Array.<String>} indexes - Index names
+   * @param {String} indexType - Index type (internal or public)
+   *
+   * @returns {Promise}
+   */
+  _addCollections (indexes, indexType) {
+    const promises = [];
+
+    for (const index of indexes) {
+      this.indexes[index] = {
+        indexType,
+        collections: []
+      };
+
+      promises.push(
+        this.internalStorage.listCollections(index)
+          .then(collections => {
+            for (const collection in collections) {
+              this.indexes[index].collections.push(collection)
+            }
+          }));
+    }
+
+    return Bluebird.all(promises);
+  }
+
+  /**
+   * Adds an index and/or a collection to the cache.
+   * An event is emitted for other cluster node synchronization if the cache
+   * has changed.
+   *
+   * @param {Object} arguments - index, collection, indexType (public)
+   *
+   * @returns {boolean} True if the cache has changed
+   */
+  add ({ index, collection=null, indexType='public' } = {}) {
     let modified = false;
 
-    if (index !== undefined) {
-      if (!this.indexes[index]) {
-        this.indexes[index] = [];
-        modified = true;
-      }
+    if (! this.indexes[index]) {
+      this.indexes[index] = {
+        indexType,
+        collections: []
+      };
 
-      if (collection && this.indexes[index].indexOf(collection) === -1) {
-        this.indexes[index].push(collection);
-        modified = true;
-      }
+      modified = true;
+    }
+
+    if (collection && this.indexes[index].collections.includes(collection)) {
+      this.indexes[index].collections.push(collection);
+
+      modified = true;
     }
 
-    if (notify && modified) {
-      this.kuzzle.emit('core:indexCache:add', {index, collection});
+    if (modified) {
+      // @todo modify cluster for indexType
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)